kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/InExpressionEvaluator.java

```diff
@@ -27,71 +27,34 @@
 
-  private static final Map<Class<? extends DataType>, BiFunction<Object, Object, Integer>>
-      COMPARATORS = createComparatorMap();
-
-  private static Map<Class<? extends DataType>, BiFunction<Object, Object, Integer>>
-      createComparatorMap() {
-    Map<Class<? extends DataType>, BiFunction<Object, Object, Integer>> map = new HashMap<>();
-    map.put(BooleanType.class, (v1, v2) -> Boolean.compare((Boolean) v1, (Boolean) v2));
-    map.put(
-        ByteType.class,
-        (v1, v2) -> Byte.compare(((Number) v1).byteValue(), ((Number) v2).byteValue()));
-    map.put(
-        ShortType.class,
-        (v1, v2) -> Short.compare(((Number) v1).shortValue(), ((Number) v2).shortValue()));
-    map.put(
-        IntegerType.class,
-        (v1, v2) -> Integer.compare(((Number) v1).intValue(), ((Number) v2).intValue()));
-    map.put(
-        DateType.class,
-        (v1, v2) -> Integer.compare(((Number) v1).intValue(), ((Number) v2).intValue()));
-    map.put(
-        LongType.class,
-        (v1, v2) -> Long.compare(((Number) v1).longValue(), ((Number) v2).longValue()));
-    map.put(
-        TimestampType.class,
-        (v1, v2) -> Long.compare(((Number) v1).longValue(), ((Number) v2).longValue()));
-    map.put(
-        TimestampNTZType.class,
-        (v1, v2) -> Long.compare(((Number) v1).longValue(), ((Number) v2).longValue()));
-    map.put(
-        FloatType.class,
-        (v1, v2) -> Float.compare(((Number) v1).floatValue(), ((Number) v2).floatValue()));
-    map.put(
-        DoubleType.class,
-        (v1, v2) -> Double.compare(((Number) v1).doubleValue(), ((Number) v2).doubleValue()));
-    map.put(
-        DecimalType.class,
-        (v1, v2) -> BIGDECIMAL_COMPARATOR.compare((BigDecimal) v1, (BigDecimal) v2));
-    map.put(StringType.class, (v1, v2) -> STRING_COMPARATOR.compare((String) v1, (String) v2));
-    map.put(BinaryType.class, (v1, v2) -> BINARY_COMPARTOR.compare((byte[]) v1, (byte[]) v2));
-    return Collections.unmodifiableMap(map);
-  }
-
   /** Validates and transforms the {@code IN} expression. */
   static In validateAndTransform(
       In in,
       Expression valueExpression,
       DataType valueDataType,
       List<Expression> inListExpressions,
       List<DataType> inListDataTypes) {
-    // TODO: [delta-io/delta#5227] Try to reuse Implicit cast and simplify comparison logic
     validateArgumentCount(in, inListExpressions);
     validateInListElementsAreLiterals(in, inListExpressions);
-    validateTypeCompatibility(in, valueDataType, inListDataTypes);
-    validateCollation(in, valueDataType, inListExpressions, inListDataTypes);
+
+    // Use implicit casting for type compatibility
+    DataType commonType = resolveCommonType(in, valueDataType, inListDataTypes);
+    Expression transformedValue = applyCastIfNeeded(valueExpression, valueDataType, commonType);
+    List<Expression> transformedList =
+        applyListCasts(inListExpressions, inListDataTypes, commonType);
+
+    validateCollation(
```

Comment:
Let's do the collation validation first, casting to common type is our implementation that may not be related to the user provided expression